src/internal/Notification.ts

```diff
@@ -134,17 +216,86 @@ export class Notification<T> {
    * @return {Notification<T>} The "error" Notification representing the
    * argument.
    * @nocollapse
+   * @deprecated remove in v8. It is NOT recommended to create instances of `Notification` directly
+   * and use them. Rather, try to create POJOs matching the signature outlined in {@link ObservableNotification}.
+   * For example: `{ kind: 'N', value: 1 }`, `{kind: 'E', error: new Error('bad') }`, or `{ kind: 'C' }`.
    */
-  static createError<T>(err?: any): Notification<T> {
-    return new Notification('E', undefined, err);
+  static createError(err?: any) {
+    return new Notification('E', undefined, err) as Notification<never> & ErrorNotification;
   }
 
   /**
    * A shortcut to create a Notification instance of the type `complete`.
    * @return {Notification<any>} The valueless "complete" Notification.
    * @nocollapse
+   * @deprecated remove in v8. It is NOT recommended to create instances of `Notification` directly
+   * and use them. Rather, try to create POJOs matching the signature outlined in {@link ObservableNotification}.
+   * For example: `{ kind: 'N', value: 1 }`, `{kind: 'E', error: new Error('bad') }`, or `{ kind: 'C' }`.
    */
-  static createComplete(): Notification<any> {
+  static createComplete(): Notification<never> & CompleteNotification {
     return Notification.completeNotification;
   }
 }
+
+/**
+ * Executes the appropriate handler on a passed `observer` given the `kind` of notification.
+ * If the handler is missing it will do nothing. Even if the notification is an error, if
+ * there is no error handler on the observer, an error will not be thrown, it will noop.
+ * @param notification The notification object to observe.
+ * @param observer The observer to notify.
+ */
+export function observeNotification<T>(notification: ObservableNotification<T>, observer: PartialObserver<T>) {
+  if (typeof notification.kind !== 'string') {
+    throw new TypeError('Invalid notification, missing "kind"');
+  }
+  switch (notification.kind) {
+    case 'N':
+      observer.next?.(notification.value!);
+      break;
+    case 'E':
+      observer.error?.(notification.error);
+      break;
+    case 'C':
+      observer.complete?.();
+      break;
+  }
+}
+
+/**
+ * A completion object optimized for memory use and created to be the
+ * same "shape" as other notifications in v8.
+ * @internal
+ */
+export const COMPLETE_NOTIFICATION = (() => createNotification('C', undefined, undefined) as CompleteNotification)();
+
+/**
+ * Internal use only. Creates an optimized error notification that is the same "shape"
+ * as other notifications.
+ * @internal
+ */
+export function errorNotification(error: any): ErrorNotification {
+  return createNotification('E', undefined, error) as any;
```

Comment:
There's one `any` left that could instead be an assertion (and the function could have an implicit return type).